### PR TITLE
Goddess statues and orbs

### PR DESCRIPTION
**Pull request by madrury**
A first draft of a boon system, which awards the player with their choice of spells after collecting enough of a "currency" item.

# Goddess Orbs
These items are randomly dropped by monsters with a fixed frequency. Here's the player after dispatching a rat, which has dropped an orb:

Screen Shot 2024-01-29 at 7 57 09 PM

When collected, they do *not* go into the player's inventory. Instead, they are added to a bag which acts as a simple counter for how many the player has collected.

Screen Shot 2024-01-29 at 8 06 17 PM

# Goddess Tiles
The orbs can be traded for magic spells at randomly spawning locations like this:

Screen Shot 2024-01-29 at 8 03 29 PM

If the player steps on the tile and has sufficient orbs (4 at the moment), a dialogue opens allowing them to learn a magic spell:

Screen Shot 2024-01-29 at 8 03 36 PM

For now, this is a static list of the same three spells.

# Enhancements?
Some ideas for elaboration on this theme:

  - Have the tiles guarded by mobs that spawn close by. In general, have them guarded by some kind of challenge.
  - Have the boons offered form a tree of sorts, with more powerful spells of the same element or theme offered once lower levels have been collected.
  - Have boons for static attributes, stats boosts and the such.

### CODE REVIEW

File: rogue/src/entity_spawners/mod.rs [hunk 1]
```diff
@@ -282,5 +290,16 @@ fn spawn_random_monster(ecs: &mut World, x: i32, y: i32, depth: i32) -> Option<M
         Some(MonsterType::BlueJelly) => monsters::blue_jelly(ecs, x, y),
         _ => {None}
     };
+    // TODO: Same as above, this should really be part of a more general drop
+    // system.
```
Comment (axiomabsolute) [inline rogue/src/entity_spawners/mod.rs:294]: "... a more general drop system."

As in something like monsters dropping items, versus just spawning them at random locations?

File: rogue/src/blessings.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                .expect("Could not insert offered blessing component.");
+            offers.push(offer);
+        }
+        None => {}
+    };
+    ecs.maintain();
+}
+
+// Add a blessing to the player's spellbook.
+pub fn receive_blessing(ecs: &mut World, blessing: Entity) {
```
Comment (madrury) [inline rogue/src/blessings.rs:44]: It's hard assumed here that all blessings are castable. To generalize, we will need to classify blessings into types and write conditional logic based on the type.

File: rogue/src/blessings.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
+
+use specs::prelude::*;
+use super::{
+    BlessingOrbBag, OfferedBlessing, BlessingSelectionTile, Position, Point,
+    WantsToDissipate, InSpellBook, Castable, entity_spawners
+};
+
+pub fn create_offered_blessings(ecs: &mut World) {
```
Comment (madrury) [inline rogue/src/blessings.rs:8]: A minimal possible implementation. Three fixed blessings.

File: rogue/src/main.rs [hunk 4]
```diff
@@ -536,13 +540,21 @@ impl GameState for State {
             }
             RunState::HazardTurn => {
                 self.run_hazard_turn_systems();
+                let get_blessing = is_player_encroaching_blessing_tile(&self.ecs)
```
Comment (axiomabsolute) [inline rogue/src/main.rs:543]: How dare you call these blessings a hazard.

I assume `RunState::HazardTurn` is being used as more of a catch-all for hazards+boons here? Is the idea that you don't want e.g. monsters to get a turn while interacting with a blessing statue?

**Review by axiomabsolute — APPROVED** [review]
🎉

File: rogue/src/components/magic.rs [hunk 5]
```diff
@@ -14,11 +14,52 @@ use serde::{Serialize, Deserialize};
...
+
+#[derive(Component, Serialize, Deserialize, Clone)]
+pub struct BlessingSelectionTile {}
+
+#[derive(Component, Serialize, Deserialize, Clone)]
+pub struct OfferedBlessing {}
+
+
+#[derive(PartialEq, Copy, Clone, Serialize, Deserialize)]
+pub enum BlessingSlot {
```
Comment (madrury) [inline rogue/src/components/magic.rs:42]: Looks like I had the idea to introduce spell slots, though I don't seem to have made much use of it yet.

File: rogue/src/blessings.rs (hunk 2 above)
Comment (axiomabsolute) [reply rogue/src/blessings.rs:44]: Would that be using something similar to `(&entities, &castables).join()` for each class of thing? I haven't dug too deeply into how that works under the covers yet, but I assume based on the mention of using a bitset that it's reasonably efficient.

Comment (madrury) [reply rogue/src/blessings.rs:44]: I haven't really thought about how to implement it, I'm honestly not quite convinced this sort of progression system is a great idea. I mostly stole the idea from Hades.

I think the conditional logic would branch on like, how the boon is "applied" to the player. Like, here, we have this `InSpellbook` component, and you attach it to the spell with a pointer to the player. If the boon was a buff, or some other static effect, we would need something comparable.

We should think about some way to model buffs, nerfs. I remember being unsatisfied with it.

File: rogue/src/entity_spawners/mod.rs (hunk 1 above)
Comment (madrury) [reply rogue/src/entity_spawners/mod.rs:294]: Yah, like dropping equipment for example.

File: rogue/src/main.rs (hunk 4 above)
Comment (madrury) [reply rogue/src/main.rs:543]: If I recall: each turn is split up into three "subturns": the player acts, monsters act, then "hazards" act. My initial hazards were like, fire which spreads and evaporates, hence the name "hazard".

I'm gonna guess I got a bit lazy here and reused the hazard subturn to run the blessing systems. This turn logic is something I'm going to need to re-learn.

**Pull request merged**